### PR TITLE
core: add isb() before reading CNTPCT/CNTVCT

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Arm ARM quite clearly mentions that such reads must be preceded by an
ISB to forbid re-ordering.

Reported-by: Olivier Deprez <Olivier.Deprez@arm.com>
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Comment by jforissier** [comment]
Why not add the `isb` inside `read_cntpct()`?

**Comment by jforissier** [comment]
Thinking a bit more about this. I find the wording in the commit description quite strong, almost implying that accessing CNTPCT or CNTVCT without ISB could never be correct which is not my understanding. Can you give a precise reference to the ARM?
My understanding is that in the vast majority of cases ISB then read is the desired behavior, but there could be exceptions.

How about adding a new inline function: `isb_read_cntvct()` for example?

**Comment by jenswi-linaro** [comment]
Reworked a bit.

File: core/arch/arm/include/kernel/delay.h [hunk 1]
```diff
@@ -44,11 +44,13 @@ static inline uint64_t arm_cnt_us2cnt(uint32_t us)
 
 static inline uint64_t timeout_init_us(uint32_t us)
 {
+	isb();
 	return read_cntpct() + arm_cnt_us2cnt(us);
```
Comment (jforissier) [inline core/arch/arm/include/kernel/delay.h:48]: Why not use `barrier_read_cntpct()` here too?

File: core/arch/arm/include/kernel/delay.h [hunk 2]
```diff
@@ -44,11 +44,13 @@ static inline uint64_t arm_cnt_us2cnt(uint32_t us)
...
 }
 
 static inline bool timeout_elapsed(uint64_t expire)
 {
+	isb();
 	return read_cntpct() > expire;
```
Comment (jforissier) [inline core/arch/arm/include/kernel/delay.h:54]: Why not use `barrier_read_cntpct()` here too?

**Review by jforissier — COMMENTED** [review]
With the below comments addressed:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: core/arch/arm/include/kernel/delay.h (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/delay.h]: Sorry, missed that.

**Comment by jenswi-linaro** [comment]
Comments addressed, tag applied.

File: core/arch/arm/kernel/tee_time_arm_cntpct.c [hunk 3]
```diff
@@ -49,11 +51,12 @@ REGISTER_TIME_SOURCE(arm_cntpct_time_source)
 
 void plat_prng_add_jitter_entropy(enum crypto_rng_src sid, unsigned int *pnum)
 {
-	uint64_t tsc = read_cntpct();
+	uint64_t tsc = 0;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/tee_time_arm_cntpct.c:54]: `uint64_t tsc = barrier_read_cntpct();` ?

File: core/arch/arm/kernel/tee_time_arm_cntpct.c [hunk 4]
```diff
@@ -16,9 +16,11 @@
 
 static TEE_Result arm_cntpct_get_sys_time(TEE_Time *time)
 {
-	uint64_t cntpct = read_cntpct();
+	uint64_t cntpct = 0;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/tee_time_arm_cntpct.c:19]: `	uint64_t cntpct = barrier_read_cntpct();` ?

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` ith minor comments on last commit.

**Comment by jenswi-linaro** [comment]
Tag applied and fixed up as requested.

**Pull request merged**